File: examples/showcase/web-backend/check.mjs
// Drive the governed Express backend and assert verdicts and attribution.
//
//     npm i -g @watchlight/sdk            (or, in this repo: cd ts && npm install && npm run build)
//     (cd examples/showcase/web-backend && npm install)   # Express — optional extra for this example only
//     node examples/showcase/web-backend/check.mjs
//
// Starts `app.mjs` on an ephemeral 127.0.0.1 port with its audit trail pointed at a
// scratch directory, sends an allowed request (alice, her account), a denied one
// (bob, the same account), a second denied one (alice, another account), three
// unauthenticated ones and a malformed one, shuts the server down, then reads the
// trail the server wrote and asserts that every decision carries the acting user
// as its principal. Exits non-zero on any failed assertion; exits 2 with a message
// if the optional web extras are not installed.
import { spawn } from "node:child_process";
import { existsSync, mkdtempSync, readFileSync, rmSync } from "node:fs";
import { createRequire } from "node:module";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { fileURLToPath } from "node:url";

const require = createRequire(import.meta.url);
const HERE = new URL(".", import.meta.url);
const APP = fileURLToPath(new URL("app.mjs", HERE));
const STARTUP_TIMEOUT_MS = 30_000;

function loadSdk() {
  for (const spec of ["@watchlight/sdk", fileURLToPath(new URL("../../../ts/dist/index.js", HERE))]) {
    try { return require(spec); } catch (e) { if (e?.code !== "MODULE_NOT_FOUND") throw e; }
  }
  throw new Error("@watchlight/sdk not found — 'npm i -g @watchlight/sdk' or build it with 'cd ts && npm run build'");
}
const { Watchlight, DENY_REASON } = loadSdk();

const TOKENS = { alice: "demo-token-alice", bob: "demo-token-bob" }; // the stand-in table in app.mjs

async function get(base, path, token) {
  const resp = await fetch(base + path, { headers: token ? { Authorization: `Bearer ${token}` } : {} });
  const text = await resp.text();
  return [resp.status, text ? JSON.parse(text) : {}];
}

/** Start the server; resolve with its base URL once it prints the port it bound. */
function startServer(auditDir) {
  const proc = spawn(process.execPath, [APP], { env: { ...process.env, WEB_BACKEND_AUDIT_DIR: auditDir }, stdio: ["ignore", "pipe", "pipe"] });
  const lines = [];
  const base = new Promise((resolve, reject) => {
    let buf = "";
    const timer = setTimeout(() => { proc.kill("SIGKILL"); reject(new Error("server did not start:\n  " + lines.join("\n  "))); }, STARTUP_TIMEOUT_MS);
    const onData = (chunk) => {
      buf += chunk;
      let nl;
      while ((nl = buf.indexOf("\n")) >= 0) {
        const line = buf.slice(0, nl); buf = buf.slice(nl + 1); lines.push(line);
        const m = /^listening on (http:\/\/127\.0\.0\.1:\d+)$/.exec(line.trim());
        if (m) { clearTimeout(timer); resolve(m[1]); }
      }
    };
    proc.stdout.setEncoding("utf8").on("data", onData);
    proc.stderr.setEncoding("utf8").on("data", onData);
    proc.on("exit", (code) => { clearTimeout(timer); reject(new Error(`server exited early (${code}):\n  ` + lines.join("\n  "))); });
  });
  const exited = new Promise((resolve) => proc.on("exit", (code, signal) => resolve({ code, signal })));
  return { proc, base, exited };
}

async function main() {
  try {
    require.resolve("express");
  } catch (e) {
    if (e?.code !== "MODULE_NOT_FOUND") throw e;
    console.error("check.mjs needs Express (optional extra, not part of @watchlight/sdk):\n" +
      "    cd examples/showcase/web-backend && npm install");
    process.exit(2);
  }

  const auditDir = mkdtempSync(join(tmpdir(), "web-backend-audit-"));
  const { proc, base: baseP, exited } = startServer(auditDir);
  let failures = 0;
  const check = (name, cond, detail = "") => {
    console.log(`  ${cond ? "✓" : "✗"} ${name}${cond || !detail ? "" : " — " + detail}`);
    if (!cond) failures++;
  };

  let allowed, deniedUser, deniedAccount, noToken, badToken, protoToken, badPath, exit;
  let forced = false; // set if the server ignored SIGTERM and had to be killed
  try {
    const base = await baseP;
    console.log(`server: ${base} (pid ${proc.pid}); audit trail → scratch directory\n`);

    // ── requests ──
    allowed = await get(base, "/accounts/acct-100/statement", TOKENS.alice);
    deniedUser = await get(base, "/accounts/acct-100/statement", TOKENS.bob);
    deniedAccount = await get(base, "/accounts/acct-200/statement", TOKENS.alice);
    noToken = await get(base, "/accounts/acct-100/statement");
    badToken = await get(base, "/accounts/acct-100/statement", "demo-token-nobody");
    protoToken = await get(base, "/accounts/acct-100/statement", "constructor"); // must not resolve via the prototype chain
    badPath = await get(base, "/accounts/acct_100!/statement", TOKENS.alice);
    for (const [label, [status, body]] of [["alice  → acct-100", allowed], ["bob    → acct-100", deniedUser],
      ["alice  → acct-200", deniedAccount], ["no token", noToken], ["unknown token", badToken], ["token 'constructor'", protoToken],
      ["malformed account id", badPath]]) {
      const keys = body && typeof body === "object" ? "[" + Object.keys(body).sort().map((k) => `'${k}'`).join(", ") + "]" : body;
      console.log(`  ${label.padEnd(22)} HTTP ${status}  ${keys}`);
    }
  } finally {
    proc.kill("SIGTERM");
    const killer = setTimeout(() => { forced = true; proc.kill("SIGKILL"); }, 10_000);
    exit = await exited;
    clearTimeout(killer);
  }

  // ── the trail the server wrote ──
  const trailPath = join(auditDir, "audit.jsonl");
  const trail = existsSync(trailPath)
    ? readFileSync(trailPath, "utf8").split("\n").filter((l) => l.trim()).map((l) => JSON.parse(l)) : [];
  rmSync(auditDir, { recursive: true, force: true });
  const decisions = trail.filter((r) => "decision" in r);
  const egress = trail.filter((r) => r.event === "egress");

  console.log("\n=== audit trail (written by the server) ===");
  for (const d of decisions) {
    console.log(`  …${(d.decision_id ?? "").slice(-6)}  ${d.decision.padEnd(5)}  principal=${d.principal}  ${d.intent}  ${d.resource}`);
  }

  console.log("\n=== assertions ===");
  const opaque = (b) => JSON.stringify(b) === JSON.stringify({ error: DENY_REASON });
  check("alice reading her account → 200 with the statement and a decision_id",
    allowed[0] === 200 && "statement" in allowed[1] && Boolean(allowed[1].decision_id));
  check("bob reading the same account → 403 with the opaque reason, no statement",
    deniedUser[0] === 403 && opaque(deniedUser[1]));
  check("alice reading another account → 403 (the policy is scoped to her account)",
    deniedAccount[0] === 403 && opaque(deniedAccount[1]));
  check("no token / unknown token / a prototype-chain name as token → 401 before any governed call",
    noToken[0] === 401 && badToken[0] === 401 && protoToken[0] === 401);
  check("a malformed account id → 400 before any governed call", badPath[0] === 400);
  check("exactly three decisions: one per authenticated request, none for the 401s and the 400",
    decisions.length === 3 && decisions.every((d) => ['User::"alice"', 'User::"bob"'].includes(d.principal)));
  const byKey = Object.fromEntries(decisions.map((d) => [`${d.principal} ${d.resource}`, d]));
  check('Allow for User::"alice" on account/acct-100', byKey['User::"alice" account/acct-100']?.decision === "Allow");
  check('Deny for User::"bob" on account/acct-100', byKey['User::"bob" account/acct-100']?.decision === "Deny");
  check('Deny for User::"alice" on account/acct-200', byKey['User::"alice" account/acct-200']?.decision === "Deny");
  check("every decision is attributed to the acting user, never to the service",
    decisions.every((d) => d.principal.startsWith('User::"') && d.principal !== d.agent));
  const allow = byKey['User::"alice" account/acct-100'] ?? {};
  check("the decision_id in alice's response is the Allow record's — the response joins the trail",
    allowed[1].decision_id === allow.decision_id);
  check("one egress record, joined to the Allow, replaced (the hook attached the decision_id)",
    egress.length === 1 && egress[0].decision_id === allow.decision_id && egress[0].replaced === true && egress[0].principal === 'User::"alice"');
  const blob = JSON.stringify(trail);
  check("the trail is value-free — no bearer token and no statement text in it",
    Object.values(TOKENS).every((t) => !blob.includes(t)) && !blob.includes("closing balance"));
  check("the server stopped on SIGTERM within 10s without being killed", !forced, `exit ${exit.code ?? exit.signal}`);

  // The policy the server loaded, executed in-process (same as `watchlight policy test`).
  const suitePath = fileURLToPath(new URL("policy.suite.json", HERE));
  const suite = JSON.parse(readFileSync(suitePath, "utf8"));
  const suiteDir = mkdtempSync(join(tmpdir(), "web-backend-suite-"));
  let report;
  try {
    report = await new Watchlight({ agent: "check", auditDir: suiteDir }).load(suitePath).test(suite.tests);
  } finally {
    rmSync(suiteDir, { recursive: true, force: true });
  }
  check(`policy.suite.json: ${report.passed}/${report.total} fixtures pass`, report.failed === 0);

  console.log(`\n${failures === 0 ? "ALL CHECKS OK" : `${failures} CHECK(S) FAILED`}`);
  process.exit(failures === 0 ? 0 : 1);
}

main().catch((e) => { console.error(e); process.exit(1); });

File: examples/showcase/web-backend/check.py
#!/usr/bin/env python3
"""Drive the governed FastAPI backend and assert verdicts and attribution.

    pip install watchlight -r examples/showcase/web-backend/requirements.txt
    python examples/showcase/web-backend/check.py

Starts `app.py` on an ephemeral 127.0.0.1 port with its audit trail pointed at a
scratch directory, sends an allowed request (alice, her account), a denied one
(bob, the same account), a second denied one (alice, another account), three
unauthenticated ones and a malformed one, shuts the server down, then reads the
trail the server wrote and asserts that every decision carries the acting user
as its principal. Exits non-zero on any failed assertion; exits 2 with a message
if the optional web extras are not installed.
"""

import importlib.util
import json
import os
import pathlib
import re
import shutil
import subprocess
import sys
import tempfile
import threading
import urllib.error
import urllib.request
from typing import Optional

from watchlight import DENY_REASON, Watchlight

HERE = pathlib.Path(__file__).parent
APP = HERE / "app.py"
STARTUP_TIMEOUT_S = 30

TOKENS = {"alice": "demo-token-alice", "bob": "demo-token-bob"}  # the stand-in table in app.py


def get(base: str, path: str, token: Optional[str] = None) -> tuple[int, dict]:
    req = urllib.request.Request(base + path, headers={"Authorization": f"Bearer {token}"} if token else {})
    try:
        with urllib.request.urlopen(req, timeout=10) as resp:
            return resp.status, json.loads(resp.read() or b"{}")
    except urllib.error.HTTPError as e:
        return e.code, json.loads(e.read() or b"{}")


def main() -> int:
    if importlib.util.find_spec("fastapi") is None or importlib.util.find_spec("uvicorn") is None:
        print("check.py needs FastAPI and uvicorn (optional extras, not part of the watchlight package):\n"
              "    pip install -r examples/showcase/web-backend/requirements.txt", file=sys.stderr)
        return 2

    audit_dir = pathlib.Path(tempfile.mkdtemp(prefix="web-backend-audit-"))
    env = {**os.environ, "WEB_BACKEND_AUDIT_DIR": str(audit_dir)}
    proc = subprocess.Popen([sys.executable, str(APP)], stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
                            text=True, env=env)
    failures = 0
    forced = False  # set if the server ignored SIGTERM and had to be killed

    def check(name: str, cond: bool, detail: str = "") -> None:
        nonlocal failures
        print(f"  {'✓' if cond else '✗'} {name}{'' if cond or not detail else ' — ' + detail}")
        failures += 0 if cond else 1

    try:
        # ── start: the app prints the port it bound ──
        killer = threading.Timer(STARTUP_TIMEOUT_S, proc.kill)
        killer.start()
        base = None
        server_lines: list[str] = []
        for line in proc.stdout:
            server_lines.append(line.rstrip("\n"))
            m = re.match(r"^listening on (http://127\.0\.0\.1:\d+)$", line.strip())
            if m:
                base = m.group(1)
                break
        killer.cancel()
        if base is None:
            print("server did not start:\n  " + "\n  ".join(server_lines), file=sys.stderr)
            return 1
        threading.Thread(target=lambda: server_lines.extend(l.rstrip("\n") for l in proc.stdout), daemon=True).start()
        print(f"server: {base} (pid {proc.pid}); audit trail → scratch directory\n")

        # ── requests ──
        allowed = get(base, "/accounts/acct-100/statement", TOKENS["alice"])
        denied_user = get(base, "/accounts/acct-100/statement", TOKENS["bob"])
        denied_account = get(base, "/accounts/acct-200/statement", TOKENS["alice"])
        no_token = get(base, "/accounts/acct-100/statement")
        bad_token = get(base, "/accounts/acct-100/statement", "demo-token-nobody")
        proto_token = get(base, "/accounts/acct-100/statement", "constructor")  # must not resolve via any prototype/attribute chain
        bad_path = get(base, "/accounts/acct_100!/statement", TOKENS["alice"])
        for label, (status, body) in [("alice  → acct-100", allowed), ("bob    → acct-100", denied_user),
                                      ("alice  → acct-200", denied_account), ("no token", no_token),
                                      ("unknown token", bad_token), ("token 'constructor'", proto_token),
                                      ("malformed account id", bad_path)]:
            print(f"  {label:22} HTTP {status}  {sorted(body) if isinstance(body, dict) else body}")
    finally:
        proc.terminate()
        try:
            proc.wait(timeout=10)
        except subprocess.TimeoutExpired:
            forced = True
            proc.kill()
            proc.wait()

    # ── the trail the server wrote ──
    trail_path = audit_dir / "audit.jsonl"
    trail = [json.loads(l) for l in trail_path.read_text(encoding="utf-8").splitlines() if l.strip()] if trail_path.exists() else []
    shutil.rmtree(audit_dir, ignore_errors=True)
    decisions = [r for r in trail if "decision" in r]
    egress = [r for r in trail if r.get("event") == "egress"]

    print("\n=== audit trail (written by the server) ===")
    for d in decisions:
        print(f"  …{d.get('decision_id', '')[-6:]}  {d['decision']:5}  principal={d['principal']}  {d['intent']}  {d['resource']}")

    print("\n=== assertions ===")
    check("alice reading her account → 200 with the statement and a decision_id",
          allowed[0] == 200 and "statement" in allowed[1] and bool(allowed[1].get("decision_id")))
    check("bob reading the same account → 403 with the opaque reason, no statement",
          denied_user[0] == 403 and denied_user[1] == {"error": DENY_REASON})
    check("alice reading another account → 403 (the policy is scoped to her account)",
          denied_account[0] == 403 and denied_account[1] == {"error": DENY_REASON})
    check("no token / unknown token / a prototype-chain name as token → 401 before any governed call",
          no_token[0] == 401 and bad_token[0] == 401 and proto_token[0] == 401)
    check("a malformed account id → 400 before any governed call", bad_path[0] == 400)
    check("exactly three decisions: one per authenticated request, none for the 401s and the 400",
          len(decisions) == 3 and all(d["principal"] in ('User::"alice"', 'User::"bob"') for d in decisions))
    by_key = {(d["principal"], d["resource"]): d for d in decisions}
    check('Allow for User::"alice" on account/acct-100',
          by_key.get(('User::"alice"', "account/acct-100"), {}).get("decision") == "Allow")
    check('Deny for User::"bob" on account/acct-100',
          by_key.get(('User::"bob"', "account/acct-100"), {}).get("decision") == "Deny")
    check('Deny for User::"alice" on account/acct-200',
          by_key.get(('User::"alice"', "account/acct-200"), {}).get("decision") == "Deny")
    check("every decision is attributed to the acting user, never to the service",
          all(d["principal"].startswith('User::"') and d["principal"] != d["agent"] for d in decisions))
    allow = by_key.get(('User::"alice"', "account/acct-100"), {})
    check("the decision_id in alice's response is the Allow record's — the response joins the trail",
          allowed[1].get("decision_id") == allow.get("decision_id"))
    check("one egress record, joined to the Allow, replaced (the hook attached the decision_id)",
          len(egress) == 1 and egress[0].get("decision_id") == allow.get("decision_id")
          and egress[0]["replaced"] is True and egress[0]["principal"] == 'User::"alice"')
    blob = json.dumps(trail)
    check("the trail is value-free — no bearer token and no statement text in it",
          not any(t in blob for t in TOKENS.values()) and "closing balance" not in blob)
    # uvicorn re-raises the signal it stopped on, so a graceful stop exits with SIGTERM's status.
    check("the server stopped on SIGTERM within 10s without being killed", not forced, f"exit {proc.returncode}")

    # The policy the server loaded, executed in-process (same as `watchlight policy test`).
    suite = json.loads((HERE / "policy.suite.json").read_text(encoding="utf-8"))
    suite_dir = tempfile.mkdtemp(prefix="web-backend-suite-")
    try:
        report = Watchlight(agent="check", audit_dir=suite_dir).load(HERE / "policy.suite.json").test(suite["tests"])
    finally:
        shutil.rmtree(suite_dir, ignore_errors=True)
    check(f"policy.suite.json: {report['passed']}/{report['total']} fixtures pass", report["failed"] == 0)

    print(f"\n{'ALL CHECKS OK' if failures == 0 else f'{failures} CHECK(S) FAILED'}")
    return 0 if failures == 0 else 1


if __name__ == "__main__":
    sys.exit(main())
